File: library/src/main/java/me/drakeet/multitype/MultiTypeAdapter.java
/*
 * Copyright 2016 drakeet. https://github.com/drakeet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.drakeet.multitype;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * @author drakeet
 */
public class MultiTypeAdapter extends RecyclerView.Adapter<ViewHolder> {

    private static final String TAG = "MultiTypeAdapter";

    private @NonNull List<?> items;
    private @NonNull TypePool typePool;
    protected @Nullable LayoutInflater inflater;


    /**
     * Constructs a MultiTypeAdapter with an empty items list.
     */
    public MultiTypeAdapter() {
        this(Collections.emptyList());
    }


    /**
     * Constructs a MultiTypeAdapter with a items list.
     *
     * @param items the items list
     */
    public MultiTypeAdapter(@NonNull List<?> items) {
        this(items, new MultiTypePool());
    }


    /**
     * Constructs a MultiTypeAdapter with a items list and an initial capacity of TypePool.
     *
     * @param items the items list
     * @param initialCapacity the initial capacity of TypePool
     */
    public MultiTypeAdapter(@NonNull List<?> items, int initialCapacity) {
        this(items, new MultiTypePool(initialCapacity));
    }


    /**
     * Constructs a MultiTypeAdapter with a items list and a TypePool.
     *
     * @param items the items list
     * @param pool the type pool
     */
    public MultiTypeAdapter(@NonNull List<?> items, @NonNull TypePool pool) {
        this.items = items;
        this.typePool = pool;
    }


    /**
     * Registers a type class and its item view binder. If you have registered the class,
     * it will override the original binder(s). Note that the method is non-thread-safe
     * so that you should not use it in concurrent operation.
     * <p>
     * Note that the method should not be called after
     * {@link RecyclerView#setAdapter(RecyclerView.Adapter)}, or you have to call the setAdapter
     * again.
     * </p>
     *
     * @param clazz the class of a item
     * @param binder the item view binder
     * @param <T> the item data type
     */
    public <T> void register(
        @NonNull Class<? extends T> clazz, @NonNull ItemViewBinder<T, ?> binder) {
        checkAndRemoveAllTypesIfNeed(clazz);
        typePool.register(clazz, binder, new DefaultLinker<T>());
    }


    /**
     * Registers a type class to multiple item view binders. If you have registered the
     * class, it will override the original binder(s). Note that the method is non-thread-safe
     * so that you should not use it in concurrent operation.
     * <p>
     * Note that the method should not be called after
     * {@link RecyclerView#setAdapter(RecyclerView.Adapter)}, or you have to call the setAdapter
     * again.
     * </p>
     *
     * @param clazz the class of a item
     * @param <T> the item data type
     * @return {@link OneToManyFlow} for setting the binders
     * @see #register(Class, ItemViewBinder)
     */
    @CheckResult
    public <T> OneToManyFlow<T> register(@NonNull Class<? extends T> clazz) {
        checkAndRemoveAllTypesIfNeed(clazz);
        return new OneToManyBuilder<>(this, clazz);
    }


    /**
     * Registers all of the contents in the specified type pool. If you have registered a
     * class, it will override the original binder(s). Note that the method is non-thread-safe
     * so that you should not use it in concurrent operation.
     * <p>
     * Note that the method should not be called after
     * {@link RecyclerView#setAdapter(RecyclerView.Adapter)}, or you have to call the setAdapter
     * again.
     * </p>
     *
     * @param pool type pool containing contents to be added to this adapter inner pool
     * @see #register(Class, ItemViewBinder)
     * @see #register(Class)
     */
    public void registerAll(@NonNull final TypePool pool) {
        for (int i = 0; i < pool.getClasses().size(); i++) {
            registerWithoutChecking(
                pool.getClasses().get(i),
                pool.getItemViewBinders().get(i),
                pool.getLinkers().get(i)
            );
        }
    }


    /**
     * Sets and updates the items atomically and safely. It is recommended to use this method
     * to update the items with a new wrapper list or consider using {@link CopyOnWriteArrayList}.
     *
     * <p>Note: If you want to refresh the list views after setting items, you should
     * call {@link RecyclerView.Adapter#notifyDataSetChanged()} by yourself.</p>
     *
     * @param items the new items list
     * @since v2.4.1
     */
    public void setItems(@NonNull List<?> items) {
        this.items = items;
    }


    @NonNull
    public List<?> getItems() {
        return items;
    }


    /**
     * Set the TypePool to hold the types and view binders.
     *
     * @param typePool the TypePool implementation
     */
    public void setTypePool(@NonNull TypePool typePool) {
        this.typePool = typePool;
    }


    @NonNull
    public TypePool getTypePool() {
        return typePool;
    }


    @Override
    public final int getItemViewType(int position) {
        Object item = items.get(position);
        return indexInTypesOf(item);
    }


    @Override
    public final ViewHolder onCreateViewHolder(ViewGroup parent, int indexViewType) {
        if (inflater == null) {
            inflater = LayoutInflater.from(parent.getContext());
        }
        ItemViewBinder<?, ?> binder = typePool.getItemViewBinders().get(indexViewType);
        binder.adapter = this;
        assert inflater != null;
        return binder.onCreateViewHolder(inflater, parent);
    }


    /**
     * This method is deprecated and unused. You should not call this method.
     * <p>
     * If you need to call the binding, use {@link RecyclerView.Adapter#onBindViewHolder(ViewHolder,
     * int, List)} instead.
     * </p>
     *
     * @param holder The ViewHolder which should be updated to represent the contents of the
     * item at the given position in the data set.
     * @param position The position of the item within the adapter's data set.
     * @throws IllegalAccessError By default.
     * @deprecated Call {@link RecyclerView.Adapter#onBindViewHolder(ViewHolder, int, List)}
     * instead.
     */
    @Override @Deprecated
    public final void onBindViewHolder(ViewHolder holder, int position) {
        throw new IllegalAccessError("You should not call this method. " +
            "Call RecyclerView.Adapter#onBindViewHolder(holder, position, payloads) instead.");
    }


    @Override @SuppressWarnings("unchecked")
    public final void onBindViewHolder(ViewHolder holder, int position, List<Object> payloads) {
        Object item = items.get(position);
        ItemViewBinder binder = typePool.getItemViewBinders().get(holder.getItemViewType());
        binder.onBindViewHolder(holder, item, payloads);
    }


    @Override
    public final int getItemCount() {
        return items.size();
    }


    int indexInTypesOf(@NonNull Object item) throws BinderNotFoundException {
        int index = typePool.firstIndexOf(item.getClass());
        if (index != -1) {
            @SuppressWarnings("unchecked")
            Linker<Object> linker = (Linker<Object>) typePool.getLinkers().get(index);
            return index + linker.index(item);
        }
        throw new BinderNotFoundException(item.getClass());
    }


    private void checkAndRemoveAllTypesIfNeed(@NonNull Class<?> clazz) {
        if (!typePool.getClasses().contains(clazz)) {
            return;
        }
        Log.w(TAG, "You have registered the " + clazz.getSimpleName() + " type. " +
            "It will override the original binder(s).");
        for (; ; ) {
            int index = typePool.getClasses().indexOf(clazz);
            if (index != -1) {
                typePool.getClasses().remove(index);
                typePool.getItemViewBinders().remove(index);
                typePool.getLinkers().remove(index);
            } else {
                break;
            }
        }
    }


    <T> void registerWithLinker(
        @NonNull Class<? extends T> clazz,
        @NonNull ItemViewBinder<T, ?> binder,
        @NonNull Linker<T> linker) {
        typePool.register(clazz, binder, linker);
    }


    /** A safe register method base on the TypePool's safety for TypePool. */
    @SuppressWarnings("unchecked")
    private void registerWithoutChecking(
        @NonNull Class clazz, @NonNull ItemViewBinder itemViewBinder, @NonNull Linker linker) {
        checkAndRemoveAllTypesIfNeed(clazz);
        typePool.register(clazz, itemViewBinder, linker);
    }
}

File: library/src/main/kotlin/com/drakeet/multitype/MultiTypeAdapter.kt
/*
 * Copyright (c) 2016-present. Drakeet Xu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.drakeet.multitype

import android.util.Log
import android.view.ViewGroup
import androidx.annotation.CheckResult
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import java.util.concurrent.CopyOnWriteArrayList
import kotlin.reflect.KClass

/**
 * @author Drakeet Xu
 */
open class MultiTypeAdapter @JvmOverloads constructor(
  /**
   * Sets and updates the items atomically and safely. It is recommended to use this method
   * to update the items with a new wrapper list or consider using [CopyOnWriteArrayList].
   *
   * Note: If you want to refresh the list views after setting items, you should
   * call [RecyclerView.Adapter.notifyDataSetChanged] by yourself.
   *
   * @since v2.4.1
   */
  open var items: List<Any> = emptyList(),
  open val initialCapacity: Int = 0,
  open var types: Types = MutableTypes(initialCapacity)
) : RecyclerView.Adapter<ViewHolder>() {

  /**
   * Registers a type class and its item view delegate. If you have registered the class,
   * it will override the original delegate(s). Note that the method is non-thread-safe
   * so that you should not use it in concurrent operation.
   *
   * Note that the method should not be called after
   * [RecyclerView.setAdapter], or you have to call the setAdapter
   * again.
   *
   * @param clazz the class of a item
   * @param delegate the item view delegate
   * @param T the item data type
   * */
  fun <T> register(clazz: Class<T>, delegate: ItemViewDelegate<T, *>) {
    unregisterAllTypesIfNeeded(clazz)
    register(Type(clazz, delegate, DefaultLinker()))
  }

  inline fun <reified T : Any> register(delegate: ItemViewDelegate<T, *>) {
    register(T::class.java, delegate)
  }

  fun <T : Any> register(clazz: KClass<T>, delegate: ItemViewDelegate<T, *>) {
    register(clazz.java, delegate)
  }

  fun <T> register(clazz: Class<T>, binder: ItemViewBinder<T, *>) {
    register(clazz, binder as ItemViewDelegate<T, *>)
  }

  inline fun <reified T : Any> register(binder: ItemViewBinder<T, *>) {
    register(binder as ItemViewDelegate<T, *>)
  }

  fun <T : Any> register(clazz: KClass<T>, binder: ItemViewBinder<T, *>) {
    register(clazz, binder as ItemViewDelegate<T, *>)
  }

  internal fun <T> register(type: Type<T>) {
    types.register(type)
    type.delegate._adapter = this
  }

  /**
   * Registers a type class to multiple item view delegates. If you have registered the
   * class, it will override the original delegate(s). Note that the method is non-thread-safe
   * so that you should not use it in concurrent operation.
   *
   * Note that the method should not be called after
   * [RecyclerView.setAdapter], or you have to call the setAdapter again.
   *
   * @param clazz the class of a item
   * @param <T> the item data type
   * @return [OneToManyFlow] for setting the delegates
   * @see [register]
   */
  @CheckResult
  fun <T> register(clazz: Class<T>): OneToManyFlow<T> {
    unregisterAllTypesIfNeeded(clazz)
    return OneToManyBuilder(this, clazz)
  }

  @CheckResult
  fun <T : Any> register(clazz: KClass<T>): OneToManyFlow<T> {
    return register(clazz.java)
  }

  /**
   * Registers all of the contents in the specified [Types]. If you have registered a
   * class, it will override the original delegate(s). Note that the method is non-thread-safe
   * so that you should not use it in concurrent operation.
   *
   * Note that the method should not be called after
   * [RecyclerView.setAdapter], or you have to call the setAdapter
   * again.
   *
   * @param types a [Types] containing contents to be added to this adapter inner [Types]
   * @see [register]
   * @see [register]
   */
  fun registerAll(types: Types) {
    val size = types.size
    for (i in 0 until size) {
      val type = types.getType<Any>(i)
      unregisterAllTypesIfNeeded(type.clazz)
      register(type)
    }
  }

  override fun getItemViewType(position: Int): Int {
    return indexInTypesOf(position, items[position])
  }

  override fun onCreateViewHolder(parent: ViewGroup, indexViewType: Int): ViewHolder {
    return types.getType<Any>(indexViewType).delegate.onCreateViewHolder(parent.context, parent)
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    onBindViewHolder(holder, position, emptyList())
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: List<Any>) {
    val item = items[position]
    getOutDelegateByViewHolder(holder).onBindViewHolder(holder, item, payloads)
  }

  override fun getItemCount(): Int = items.size

  /**
   * Called to return the stable ID for the item, and passes the event to its associated delegate.
   *
   * @param position Adapter position to query
   * @return the stable ID of the item at position
   * @see ItemViewDelegate.getItemId
   * @see RecyclerView.Adapter.setHasStableIds
   * @since v3.2.0
   */
  override fun getItemId(position: Int): Long {
    val item = items[position]
    val itemViewType = getItemViewType(position)
    return types.getType<Any>(itemViewType).delegate.getItemId(item)
  }

  /**
   * Called when a view created by this adapter has been recycled, and passes the event to its
   * associated delegate.
   *
   * @param holder The ViewHolder for the view being recycled
   * @see RecyclerView.Adapter.onViewRecycled
   * @see ItemViewDelegate.onViewRecycled
   */
  override fun onViewRecycled(holder: ViewHolder) {
    getOutDelegateByViewHolder(holder).onViewRecycled(holder)
  }

  /**
   * Called by the RecyclerView if a ViewHolder created by this Adapter cannot be recycled
   * due to its transient state, and passes the event to its associated item view delegate.
   *
   * @param holder The ViewHolder containing the View that could not be recycled due to its
   * transient state.
   * @return True if the View should be recycled, false otherwise. Note that if this method
   * returns `true`, RecyclerView *will ignore* the transient state of
   * the View and recycle it regardless. If this method returns `false`,
   * RecyclerView will check the View's transient state again before giving a final decision.
   * Default implementation returns false.
   * @see RecyclerView.Adapter.onFailedToRecycleView
   * @see ItemViewDelegate.onFailedToRecycleView
   */
  override fun onFailedToRecycleView(holder: ViewHolder): Boolean {
    return getOutDelegateByViewHolder(holder).onFailedToRecycleView(holder)
  }

  /**
   * Called when a view created by this adapter has been attached to a window, and passes the
   * event to its associated item view delegate.
   *
   * @param holder Holder of the view being attached
   * @see RecyclerView.Adapter.onViewAttachedToWindow
   * @see ItemViewDelegate.onViewAttachedToWindow
   */
  override fun onViewAttachedToWindow(holder: ViewHolder) {
    getOutDelegateByViewHolder(holder).onViewAttachedToWindow(holder)
  }

  /**
   * Called when a view created by this adapter has been detached from its window, and passes
   * the event to its associated item view delegate.
   *
   * @param holder Holder of the view being detached
   * @see RecyclerView.Adapter.onViewDetachedFromWindow
   * @see ItemViewDelegate.onViewDetachedFromWindow
   */
  override fun onViewDetachedFromWindow(holder: ViewHolder) {
    getOutDelegateByViewHolder(holder).onViewDetachedFromWindow(holder)
  }

  private fun getOutDelegateByViewHolder(holder: ViewHolder): ItemViewDelegate<Any, ViewHolder> {
    @Suppress("UNCHECKED_CAST")
    return types.getType<Any>(holder.itemViewType).delegate as ItemViewDelegate<Any, ViewHolder>
  }

  @Throws(DelegateNotFoundException::class)
  internal fun indexInTypesOf(position: Int, item: Any): Int {
    val index = types.firstIndexOf(item.javaClass)
    if (index != -1) {
      val linker = types.getType<Any>(index).linker
      return index + linker.index(position, item)
    }
    throw DelegateNotFoundException(item.javaClass)
  }

  private fun unregisterAllTypesIfNeeded(clazz: Class<*>) {
    if (types.unregister(clazz)) {
      Log.w(TAG, "The type ${clazz.simpleName} you originally registered is now overwritten.")
    }
  }

  companion object {
    private const val TAG = "MultiTypeAdapter"
  }
}
